### ISSUE TITLE
Override rules from `parent_config`

### ISSUE THREAD

**Issue by sindresorhus** [realm/SwiftLint#3469]
I have the config:

```yml
parent_config: https://github.com/sindresorhus/swiftlint-config/raw/723a1fa/.swiftlint.yml
```

In the parent config, I use `only_rules` to define the rules I want. I want to reuse the https://github.com/sindresorhus/swiftlint-config/raw/723a1fa/.swiftlint.yml config in many projects, but in some projects, I might want to add some extra rules, and maybe disable some. This is currently not possible. I think `only_rules` in a child config should be able to extend a `only_rules` in a parent config. This is possible in any other linter, like ESLint in the JavaScript ecosystem.

```yml
parent_config: https://github.com/sindresorhus/swiftlint-config/raw/723a1fa/.swiftlint.yml
only_rules:
  - attributes
disabled_rules:
  - todo
```

Results in:

> SwiftLint Configuration Error: 'disabled_rules' or 'opt_in_rules' cannot be used in combination with 'only_rules' – Falling back to default configuration

---

// @fredpi

**Comment by fredpi** [comment]
@sindresorhus Thanks for reporting! Will look into this tomorrow

**Comment by galu08** [comment]
Hi! I was going to open an issue asking something that I think is similar to this.
I don't understand how child/parent configuration works. 

ie:
I want to use a base configuration file with some rules for the main project `.swiftlint.yml` and I want to disable some rules from it to use only for tests. So, I created a child config inside the test folder `.swiftlint-tests.yml`.

**.swiftlint.yml:**
```
child_config: MyTests/.swiftlint-test-base.yml
...
```

**.swiftlint-tests.yml:**
``` 
disabled_rules:
- function_body_length 
- type_body_length 
- large_tuple 
```

**File structure:**
```
MyProject
    |_ .swiftlint.yml
    |_ Source
    |_ MyTests
        |_ .swiftlint-tests.yml
```

Is this possible ? 
In my tests I got that the disabled rules are disabled for the whole project. Not only for the directory where it's placed the child file.

I'm using swiftLint version 0.42.0

Thanks!!

**Comment by fredpi** [comment]
@sindresorhus In fact, you can override the `only_rules` from the parent configuration, but only with the following structure:

```yaml
parent_config: https://github.com/sindresorhus/swiftlint-config/raw/723a1fa/.swiftlint.yml
opt_in_rules:
  - attributes
disabled_rules:
  - todo
```

The difference is that `opt_in_rules` is used instead of `only_rules`.

I admit that's super confusing, but there's a reason for that: Imagine that at some point in time, you would remove the `disabled_rules` key because you only want to extend upon the `only_rules` from the parent config (without disabling any). Now, SwiftLint would interpret your child configuration totally different, and in fact ignore all the paren't s configs `only_rules` because now the child config specifies the `only_rules`. Having the `only_rules` key semantics change depending on the existence of the `disabled_rules` key doesn't make sense to me. As the mode (optIn/disabled vs. only rules) is only deduced from the keys used in the configuration file, I also can't think of a better way to do this.

That being said, I think this behavior should at least be documented in the README, because this is probably a frequent use case. Would you mind creating a PR adjusting the README so that it explains how to achieve your desired functionality (by using the `opt_in_rules` key), @sindresorhus?

We could also update the warning given by SwiftLint to something like this:

> SwiftLint Configuration Error: 'disabled_rules' or 'opt_in_rules' cannot be used in combination with 'only_rules'. If you are trying to refine the 'only_rules' from a parent configuration, use 'opt_in_rules' instead of 'only_rules'  – Falling back to default configuration

What do you think of the latter, @jpsim? I don't have strong opinions on whether such a warning is worth it.

**Comment by fredpi** [comment]
@galu08 Your issue is unrelated to this issue, but as it seems simple to answer, I'll do it here. If the answer doesn't work for you, please open a separate issue and mention me.

What you are looking for is [nested configurations](https://github.com/realm/SwiftLint/#nested-configurations). To use it do the following:

- Just use your main `.swiftlint.yml` as you would normally. You **mustn't** use the `child_config` key, as that has an effect on the whole directory and not only on the directory the child config is placed in.
- Rename `MyTests/.swiftlint-tests.yml` to `MyTests/.swiftlint.yml`
- Run SwiftLint as you would normally, but make sure that you **don't** explicitly specify the `--config` parameter, because that disables nested configurations.

**Comment by sindresorhus** [comment]
I think the current way the config is structured is quite limiting and confusing. I have a [lot of experience](https://github.com/xojs/xo) with linter config from the JavaScript ecosystem. Here's how I would do it:

- All rules are disabled by default.

- Only one way to enable/disable individual rules:
```yaml
rules:
  attributes: true
  todo: false
```
This makes it much easier to extend parent configs, as individual entries in the `rules` dictionary can override the parent config.

- Recommended preset, which would be what the current defaults are.
```yaml
presets:
  builtin: true
```

**Comment by fredpi** [comment]
@sindresorhus Thanks for your suggestion!

Initially, the SwiftLint configuration system had not been designed with nested / child / parent configs in mind as those features only came later (nested configs) or even just recently (parent / child configs). That being said, when starting afresh, maybe the design you suggested would be a better (and more common) approach. But at this point, I don't think it's a good idea to change the configuration system. Here's why:

- Changing the way configurations work would break SwiftLint for almost any user and cause much irritation.
- Currently, when no configuration file is specified, SwiftLint takes the preset: All rules that aren't opt-in-rules. That doesn't change when an empty configuration file is added, which I think is reasonable behavior: For me, an empty configuration has a meaning like "No changes – just use the default settings". With the approach you suggested, adding `presets: builtin: true` would be needed in the configuration file to keep the preset functionality. In contrast, an empty configuration file would have different semantics (no activated rules).
- I don't think the current system is as bad as it may seem with the specific problem raised in this issue. While maybe the differentiation between the two modes `only_rules` and `optIn/disabled` isn't as effective as the approach you suggested (which in fact is quite effective), it's still expressive, works for most users and doesn't cause regular irritation as far as I know. And the more I think of it, the less irritating the problem described in this issue seems to me: We're just dealing with a config that opts in to some rules and disables some rules based on the preset from a parent config, not caring at all what mode the parent config is using.
- If I understand it correctly, what your suggested approach can't do is take a parent config, but then completely discard the rules activated by that config by using something like `only_rules`. Now, one may wonder why one would specify a parent config when the rule activations of the parent config get overridden in their entirety, but there's a reason: Custom Rule Inheritance.
- Changing the configuration system, considering all possible use cases, documenting it and adding extensive tests is quite a lot of work and at least I wouldn't have the time & commitment to do it anytime soon.

Of course this is just my opinion and maybe I oversaw important aspects with my argumentation. But at least the last point I mentioned is a fact ;)

Maybe also some more experienced contributors like @jpsim want to state their opinion on this issue?

**Comment by lo1tuma** [comment]
> Changing the way configurations work would break SwiftLint for almost any user and cause much irritation.

I don’t think that has to be the case. For a smooth transition SwiftLint could support both config formats, the legacy format and the new format. The new format could be an opt-in feature, which could be enabled e.g. per CLI argument (`swiftlint --use-new-config-format`).
